File: DataStructures/BynaryTree/Python/btree.py
"""
btree.py

Author: Simone Pellacani (pellacanisimone2017@gmail.com)
Brief: A module implementing a binary tree with insertion, deletion, search, and balancing functionalities.
Version: 0.1
Date: 2024-11-08

Copyright (c) 2024

This module provides an implementation for a binary tree structure with various functionalities, including node insertion, in-order traversal, balancing, and deletion. Each function is designed to operate on nodes and ensure efficient handling of binary tree properties.

Classes:
    - BTreeMeta: A meta-class that enhances the BTree with mathematical utilities, such as log base 2 calculations and height estimation.
    - BTree: A binary tree structure that supports insertion, deletion, traversal, rebalancing, and searching for elements.

Functions:
    - insert(value): Inserts a node with the specified value.
    - delete(key): Deletes a node with the specified key and rebalances the tree if necessary.
    - search(key): Searches for a node containing the specified key.
    - rebalance(): Rebalances the tree to ensure optimal height.
    - __inOrderTraversal(): Returns the tree elements in in-order traversal.

Example:
    ```python
    # Example of how to use the BTree class
    tree = BTree()
    tree.insert(10)
    tree.insert(5)
    tree.insert(15)
    tree.rebalance()
    node = tree.search(5)
    print(node)  # Prints the node with value 5 if found
    tree.delete(10)
    ```

"""

from btnode import BNode

class BTreeMeta(type):
    """Mathematical Meta-Class for binary tree utilities such as height calculations."""
    
    def __new__(cls, name, bases, class_dict):
        def logBase2(self, value: int) -> int:
            """Calculates the integer log base 2 of a value."""
            log = 1
            if value <= 0:
                return 0
            while value > 1:
                value >>= 1
                log += 1
            return log

        class_dict['logBase2'] = logBase2
        class_dict['calculateHeight'] = lambda self: int(self.logBase2(self._size + 1)) - 1
        return super().__new__(cls, name, bases, class_dict)

class BTree(metaclass=BTreeMeta):
    """Binary Tree class supporting insertion, deletion, search, and balancing."""
    
    def __resetInsertionQueue(self):
        """Initializes the insertion queue for faster insertion handling."""
        self.__insertionQueue = [self.root] if self.root else []

    def __incrementNodeCount(self):
        """Increments node count and recalculates tree height."""
        self._size += 1
        self.__height = self.calculateHeight()

    def __initRoot(self, rootValue: int):
        """Initializes the root node with the specified value."""
        self.root = BNode(rootValue) if bool(rootValue) else None
        self.__type = type(rootValue) if bool(rootValue) else None  # Compatibility tag
        self.__resetInsertionQueue()
        self._size = 0
        self.__height = 0
        if bool(rootValue):
            self.__incrementNodeCount()

    def __init__(self, rootValue=None):
        """Constructor for BTree, optionally initializing with a root value."""
        self.__initRoot(rootValue)

    def __eq__(self, b1: BNode, b2: BNode):
        """Checks equality between two nodes."""
        return b1 == b2

    def __str__(self):
        """Returns a string representation of the tree."""
        def buildString(node):
            if node is None:
                return ""
            leftStr = buildString(node.getLeft())
            rightStr = buildString(node.getRight())
            return f"({node}{leftStr}{rightStr})"
        return buildString(self.root) if self.root else ""

    def __updateQueue(self, current):
        """Adds both child nodes of the current node to the insertion queue if present."""
        self.__insertionQueue.extend(current.getBoth())

    def __setFirstFree(self, current, newNode):
        """Inserts a new node as a child of the current node based on availability."""
        if current.hasOneChild():
            newNode.setParent(current)
            current.setRight(newNode)
            self.__updateQueue(current)
            return True
        elif current.isLeaf():
            newNode.setParent(current)
            current.setLeft(newNode)
            self.__insertionQueue.insert(0, current)
            return True
        else:
            raise Exception('Unexpected error in node placement')

    def insert(self, value):
        """Inserts a node with the specified value into the tree."""
        if self.__type and not isinstance(value, self.__type):
            raise TypeError(f"Incompatible value type: expected {self.__type.__name__}, got {type(value).__name__}")
        if bool(value):
            if self.root is None:
                self.__initRoot(value)
            else:
                newNode = BNode(value)
                while self.__insertionQueue and newNode:
                    current = self.__insertionQueue.pop(0)
                    if current.hasBothChildren():
                        self.__updateQueue(current)
                    elif self.__setFirstFree(current, newNode):
                        newNode = None
                self.__incrementNodeCount()

    def __inOrderTraversal(self):
        """Performs an in-order traversal of the tree and returns elements in a list."""
        def traverse(node):
            return traverse(node.getLeft()) + [node.getInfo()] + traverse(node.getRight()) if node else []
        return traverse(self.root)

    def __insertRebalance(self):
        """Rebuilds the tree to be balanced based on in-order traversal elements."""
        elements = self.__inOrderTraversal()
        self.root = None
        for elem in elements:
            self.insert(elem)

    def rebalance(self):
        """Public method to rebalance the tree."""
        self.__insertRebalance()

    def search(self, key):
        """Searches for a node with the specified key in the tree."""
        queue = [self.root]
        while queue:
            node = queue.pop(0)
            if node is None:
                continue
            if node.getInfo() == key:
                return node
            queue.extend(node.getBoth())
        return None

    def delete(self, key):
        """Deletes the specified key from the tree and rebalances."""
        elements = [x for x in self.__inOrderTraversal() if x != key]
        self._size = len(elements)
        self.__height = self.calculateHeight()
        self.__initRoot(None)
        for elem in elements:
            self.insert(elem)

File: DataStructures/BynaryTree/C/btree.c
/**
 * @file btree.c
 * @author Simone Pellacani (pellacanisimone2017@gmail.com)
 * @brief This module provides an implementation for a binary tree structure with various 
 * functionalities, including node insertion, in-order traversal, balancing, and deletion. 
 * Each function is designed to operate on nodes and ensure efficient 
 * handling of binary tree properties.
 * 
 * @version 0.1
 * @date 2024-11-08
 * 
 * @copyright Copyright (c) 2024
 * 
 */


#include <stdio.h> // #include <iostream>
#include "btree.h"
#include "deque.h"



static int logBase2(int value) {
  int log = 1;
  if (value <= 0) { return 0; }
  while (value > 1) {
    value >>= 1;  
    log += 1;
  }
  return log;
}

static int calculateHeight(int size) { return logBase2(size + 1) - 1; }

static void incrementNodeCount(btree *t) {
  t->size++;
  t->height = calculateHeight(t->size);
}

btree* newTree(bnode* n) {
  btree* r = (btree*)malloc(sizeof(btree));
  r->root = n;
  r->size = (n == NULL) ? 0 : 1;
  r->height = calculateHeight(r->size);
  r->insertionQueue = newQueue();
  if (n != NULL) enqueue(r->insertionQueue, n);
  return r;
}


static bool same_recoursive(bnode * b1, bnode * b2) {
  if (b1 == NULL || b2 == NULL) 
    return (b1 == NULL && b2 == NULL);
  else {
    bool left = same_recoursive(b1->left, b2->left);
    bool right = same_recoursive(b1->right, b2->right);
    return b1->inf == b2->inf && left && right;
  }
}

bnode* getRoot(btree *t) { return t->root; }

void insertRoot(btree *t, bnode* n) {
  if (getRoot(t) != NULL) 
    printf("Error: root already exists.\n");
  else
    t->root = n;
}

bool same(btree * b1, btree * b2) {
  return same_recoursive(getRoot(b1) , getRoot(b2)  );
}



static bool setFirstFree(btree *t, bnode *current, bnode *newNode) {
  if ( hasOneChild(current)) { // empty right child
    newNode->parent = current;
    current->right = newNode;
    enqueue(t->insertionQueue,getLeft(current));
    enqueue(t->insertionQueue,getRight(current));
    return true;
  } 
  else if (isLeaf(current) ) { // empty left child
    newNode->parent = current;
    current->left = newNode;
    push(t->insertionQueue, current);
    return true;
  }
  return false;
}


void insert(btree *t, tipo_inf value) {
  bnode* newNode = getNewBTNode(value);

  if (getRoot(t) == NULL) {
    insertRoot(t, newNode);
    incrementNodeCount(t);
    enqueue(t->insertionQueue, newNode);
  } 
  else {
    bool inserted = false;
    while (!inserted) {
      bnode* current = dequeue(t->insertionQueue);
      if (hasBothChildren(current)) { // both not NULL
        enqueue(t->insertionQueue,getLeft(current));
        enqueue(t->insertionQueue,getRight(current));
      }
      if (setFirstFree(t, current, newNode)) {
        inserted = true;
        incrementNodeCount(t);
      } 
    }
  }
}



void freeTree(bnode *node) {
  if (node != NULL) {
    freeTree(node->left);
    freeTree(node->right);
    free(node);  
  }
}

void fullTreeDelete(btree * t) {
  freeQueue(t->insertionQueue);
  freeTree(t->root) ;
  free(t);
}

static bnode* searchNode(bnode* n, tipo_inf key) {
  if (n == NULL) return NULL;
  if (compare(getInfo(n), key) == 0) return n;

  bnode* left_result = searchNode(getLeft(n), key);
  if (left_result != NULL) return left_result;
  
  return searchNode(getRight(n), key);
}

bnode* search(btree * t, tipo_inf key) { return searchNode(getRoot(t) , key);}

static void inOrderTraversal(bnode* node, tipo_inf * arr, int * index) {
  if (node == NULL) return;
  if (node->left  != NULL)
    inOrderTraversal(node->left,  arr,  index );
  arr[index[0]++] = node->inf;
  if (node->right != NULL)
    inOrderTraversal(node->right, arr,  index );
}


static btree* createTreeFromValues(tipo_inf* values, int * size, bool toDelete, tipo_inf delKey) {
  btree* tree = NULL;
  bool first = true;
  for (int i = 0; i < size[0]; i++) {
    if  ( toDelete && compare(values[i], delKey) == 0 ) {
      toDelete=false; continue;
    }
    if (first) {
      tree = newTree(getNewBTNode(values[i]));
      first = false;
    } else {
      insert(tree, values[i]);
    }
  }
  return tree;
}


btree * rebalance(btree * tree){
  if (tree == NULL || getRoot(tree) == NULL) return NULL;
  tipo_inf* values = (tipo_inf*) malloc(tree->size * sizeof(tipo_inf));
  int index[] = {0};
  inOrderTraversal(tree->root, values, index);
  fullTreeDelete(tree);
  tipo_inf app=0;
  tree = createTreeFromValues(values,index,false,app); // false app, not delete
  free(values);
  return tree;
}


btree * deleteNode(btree *tree, tipo_inf key) {
  if (tree == NULL || getRoot(tree) == NULL) return false;
  bnode *to_delete = search(tree, key);
  if (to_delete == NULL ) return false;
  tipo_inf* values = (tipo_inf*) malloc(tree->size * sizeof(tipo_inf));
  int index[] = {0};
  inOrderTraversal(tree->root, values, index);
  fullTreeDelete(tree);
  tree = createTreeFromValues(values,index,true,key); // true delete
  free(values);
  return tree;
}

static void  printPreOrderRecursive(bnode *node) {
  if (node == NULL) { return; }
  else {
    printf("(");  
    if (getLeft(node) !=  NULL)
      printPreOrderRecursive(getLeft(node)); 
    print(&node->inf);  
    if (getRight(node) !=  NULL )
      printPreOrderRecursive(getRight(node)); 
    printf(")");
  }
}
void printPreOrder(btree *tree) {
  if (tree == NULL || tree->root == NULL) {
    printf("L'albero è vuoto.\n");
    return;
  }
  printPreOrderRecursive(tree->root);
  printf("\n");  
} 
